### PR TITLE
[google_maps] getCameraPosition

### PR DESCRIPTION
**Pull request by KlausJokisuo**
## Description

*Adds method to get current camera position via controller. Currently Android Only*

## Related Issues

*flutter/flutter#34161*

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [ ] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [ ] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

### CODE REVIEW

**Comment by iskakaushik** [comment]
We currently require changes to have both iOS and android impl. I can help you with the iOS part but it's going to take a little bit before I can get to this. I would also appreciate any help from the community with the iOS part of this PR to have this be merged faster.

**Comment by KlausJokinen** [comment]
> We currently require changes to have both iOS and android impl. I can help you with the iOS part but it's going to take a little bit before I can get to this. I would also appreciate any help from the community with the iOS part of this PR to have this be merged faster.

I understand. I hope someone can pitch in and do the iOS part. Hopefully in the future I have access to Mac.

**Comment by SirJohn2024** [comment]
Tested on iOS, works ok also on Android... 
If you're on Windows you should be able to run a macOS VM to test these things... That's what I do, eventually :-)
...
  } else if ([call.method isEqualToString:@"camera#animate"]) {
    [self animateWithCameraUpdate:ToCameraUpdate(call.arguments[@"cameraUpdate"])];
    result(nil);
  } else if ([call.method isEqualToString:@"map#getCameraPosition"]) {
    if (_mapView != nil) {
      result(PositionToJson(_mapView.camera));
    } else {
      result([FlutterError errorWithCode:@"GoogleMap uninitialized"
                                 message:@"getVisibleRegion called prior to map initialization"
                                 details:nil]);
    }
  } else if ([call.method isEqualToString:@"camera#move"]) {
    [self moveWithCameraUpdate:ToCameraUpdate(call.arguments[@"cameraUpdate"])];
    result(nil);
...

**Comment by KlausJokinen** [comment]
> Tested on iOS, works ok also on Android...
> If you're on Windows you should be able to run a macOS VM to test these things... That's what I do, eventually :-)
> ...
> `
> } else if ([call.method isEqualToString:@"map#getCameraPosition"]) {
> 
> ```
> if (_mapView != nil) {
> 
>   result(PositionToJson(_mapView.camera));
> 
> } else {
> 
>   result([FlutterError errorWithCode:@"GoogleMap uninitialized"
> 
>                              message:@"getVisibleRegion called prior to map initialization"
> 
>                              details:nil]);
> 
> }
> ```
> 
> } else if ([call.method isEqualToString:@"camera#move"]) {
> 
> `
> ...

Can you fix formatting and tell me what file you modified. Thanks 👍

**Comment by SirJohn2024** [comment]
Add the following to ios\Classes\GoogleMapController.m in the onMethodCall() method, any place should be ok since it's just a sequence of `else if{}`:

```
  } else if ([call.method isEqualToString:@"map#getCameraPosition"]) {
    if (_mapView != nil) {
      result(PositionToJson([self cameraPosition]));
    } else {
      result([FlutterError errorWithCode:@"GoogleMap uninitialized"
                                 message:@"getVisibleRegion called prior to map initialization"
                                 details:nil]);
    }
  } else...

```

**Comment by KlausJokinen** [comment]
Pushed commit with iOS side. For some reason the tests are failing. Ideas? @iskakaushik

**Comment by SirJohn2024** [comment]
Seems to me a format problem, the last else has a double indentation:
```
    } else {
    result(FlutterMethodNotImplemented);
  }
}
```
should be:

```
  } else {
    result(FlutterMethodNotImplemented);
  }
}
```

Cirrus is very picky about formatting, I wasn't able to find a Google certified formatter for ObjectiveC/Swift code as the google-java-format plugin...  :-/

**Comment by rafael-mq** [comment]
Hello guys, I'm trying to use this on a project in order to get the map bearing on the "onCameraIdle". Any chances of it getting to stable. Is there anything I can do to help?

File: packages/google_maps_flutter/ios/Classes/GoogleMapController.m [hunk 1]
```diff
@@ -240,8 +240,16 @@ - (void)onMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
...
       result(@[ @(NO), error ]);
     }
-  } else {
-    result(FlutterMethodNotImplemented);
+  } else if ([call.method isEqualToString:@"map#getCameraPosition"]) {
+      if (_mapView != nil) {
+        result(PositionToJson([self cameraPosition]));
+      } else {
+        result([FlutterError errorWithCode:@"GoogleMap uninitialized"
+                                   message:@"getVisibleRegion called prior to map initialization"
```
Comment (stuartmorgan) [inline packages/google_maps_flutter/ios/Classes/GoogleMapController.m:248]: This should be `getCameraPosition`

**Review by stuartmorgan — COMMENTED** [review]
It looks like this was just never removed from the backlog; it should be straightforward to review once it's completed.

Per the checklist in the PR template, this will need:
- tests
- version and CHANGELOG updates

The package has also been partially federated since the PR was written, so will need a secondary PR to check the platform interface package; see https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changing-federated-plugins for details.

I'm going to mark it as a draft since it's not ready for review in the current form, but once it's ready please remove the draft designation and we'll get it reviewed!

**Comment by Hixie** [comment]
@KlausJokisuo Are you still interested in working on this PR?

**Comment by KlausJokisuo** [comment]
@Hixie Hi there, I simply don't have time at the moment. I'll close this one.

**Pull request closed without merging**